h2o-algos/src/main/java/hex/tree/gbm/GBM.java

```diff
@@ -246,8 +248,69 @@ protected Frame makeValidWorkspace() {
         new FillVecWithConstant(_initialPrediction)
                 .doAll(vec_tree(_train, 0), _parms._build_tree_one_node);  // Only setting tree-column 0
       }
+
+      // Mark all rows that either have zero weights or the label is NA as decided
+      // This is way we can avoid processing them when building trees
+      final long zeroWeights = hasWeightCol() ? _weights.length() - _weights.nzCnt() : 0; 
+      if (zeroWeights > 0 ||                // has some zero weights or 
+              response().naCnt() > 0) {     // there are NAs in the response
+        Vec[] vecs = new Vec[]{_response};
+        if (hasWeightCol())
+          vecs = ArrayUtils.append(vecs, _weights);
+        for (int k = 0; k < _nclass; k++) {
+          Vec nidsVecK = vec_nids(_train, k);
+          if (nidsVecK.min() == DECIDED_ROW) {
+            // classes not present in the training frame are skipped
+            assert _model._output._distribution[k] == 0;
+            assert nidsVecK.isConst();
+            continue;
+          }
+          vecs = ArrayUtils.append(vecs, nidsVecK);
+        }
+        _skippedCnt = new MarkDecidedRows().doAll(vecs).markedCnt;
+        assert _skippedCnt >= zeroWeights;
+        assert _skippedCnt >= response().naCnt();
+        assert _skippedCnt <= response().naCnt() + zeroWeights;
+      } else
+        _skippedCnt = 0;
     }
 
+    class MarkDecidedRows extends MRTask<MarkDecidedRows> {
+      int markedCnt;
+      @Override
+      public void map(Chunk[] cs) {
+        final int colStart;
+        final Chunk y;
+        final Chunk weights;
+        if (hasWeightCol()) {
+          y = cs[0];
+          weights = cs[1];
+          colStart = 2;
+        } else {
+          y = cs[0];
+          weights = new C0DChunk(1, cs[0]._len);
```

Comment:
Is this chunk initialization necessary? You can edit the condition to ask first if `weights != null` or call `hasWeightColumn()` method instead.. Or am I missing something?

Reply:
This is a trick used oftentimes in the code at other places to avoid checking for null. It is fast and simplifies the flow of the code.

Reply:
Thanks for the explanation. :+1: